stock_available_to_promise_release/models/stock_move.py

```diff
@@ -616,27 +700,74 @@ def unrelease(self, safe_unrelease=False):
+                todo_origin_moves = origin_moves.filtered(
                     lambda m: m.state not in ("done", "cancel")
                 )
-                if origin_moves:
-                    origin_moves = move._split_origins(origin_moves)
-                    impacted_picking_ids.update(origin_moves.mapped("picking_id").ids)
+                qty_cancelled_at_step = 0
+                if todo_origin_moves:
+                    qty_to_split = qty_to_unrelease - qty_returned_for_move
+                    moves_to_cancel = move._split_origins(
+                        todo_origin_moves, qty=qty_to_split
+                    )
+                    impacted_picking_ids.update(moves_to_cancel.picking_id.ids)
                     # avoid to propagate cancel to the original move
-                    origin_moves.write({"propagate_cancel": False})
-                    # origin_moves._action_cancel()
-                    moves_to_cancel |= origin_moves
-            moves_to_cancel._action_cancel()
+                    moves_to_cancel.write({"propagate_cancel": False})
+                    moves_to_cancel_for_move |= moves_to_cancel
+                    qty_cancelled_at_step = sum(moves_to_cancel.mapped("product_qty"))
+                # checking that for the current step (pick/pack/ship)
+                # move.product_uom_qty == step.cancelled_qty + move.returned_quanty
+                # If not the case, we have to move back goods in stock.
+                qty_to_return_at_step = (
+                    qty_to_unrelease - qty_cancelled_at_step - qty_returned_for_move
+                )
+                done_moves = origin_moves.filtered(lambda m: m.state == "done")
+                if float_is_zero(qty_to_return_at_step, precision_rounding=rounding):
+                    continue
+                elif not move.rule_id.allow_unrelease_return_done_move:
+                    # Without allow_unrelease_return_done_move enabled,
+                    # only moves that aren't done can be unreleased.
+                    msg_args = {
+                        "move_name": move.name,
+                        "done_move_names": ", ".join(done_moves.mapped("name")),
+                    }
+                    message = _(
+                        (
+                            "You cannot unrelease the move %(move_name)s "
+                            "because some origin moves %(done_move_names)s are done"
+                        ),
+                        **msg_args
+                    )
+                    raise UserError(message)
+                # Multiple pickings can satisfy a move
+                # -> len(move.move_orig_ids.picking_id) > 1
+                # Group done_moves per picking, and create returns
+                returneable_qty = done_moves._get_returneable_qty_per_move(
+                    qty_to_return_at_step, qty_to_return_per_move
+                )
+                qty_returned_for_move += returneable_qty
```

Comment:
I would rename `_get_...` method name as it is not getting anything but more setting something
```suggestion
                done_origin_moves._unrelease_set_qty_to_return_per_move(
                    qty_to_return(_at_step), qty_to_return_per_move
                )
                qty_returned_for_move += qty_to_return(_at_step)
```

Reply:
your version isn't accurate, as it is possible that a current step `qty_to_return(_at_step) > sum(done_origin_moves.mapped(product_qty))`
In this case, `returneable_qty < qty_to_return(_at_step)` and the difference would be returned at next iteration.

The main goal is to get `returneable_qty` in order to keep `qty_returned_for_move` up to date, but qty_to_return isn't right.
I don't mind the naming of the method, though.